### PR TITLE
base-files: fix wrong ucidef_set_network_device_mac network-device entry

### PR DESCRIPTION
**Pull request by MadnessASAP**
The ucidef_set_network_device_* functions in uci-defaults.sh disagree on whether to use "network-device" or "network_device" in board.json. With the additional caveat that jshn will translate hyphens (-) into underscores (_). This casues problems in netifd which currently expects "network-device" causing boards which depend on assigning MACs in board.json via uci-defaults.sh (or jshn in general) to fail.

This commit addresses the issue by using network_device in uci-defaults.sh and by patching netifd to also use network_device with a fallback to network-device if nothing is returned. The patch to netifd will be/has already been submitted to upstream hopefully making the patch here irreleveant in the near future.

The bug was uncovered in the forums here:
https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875/2596

### CODE REVIEW

**Comment by MadnessASAP** [comment]
https://github.com/openwrt/netifd/pull/13

**Comment by Borromini** [comment]
To be sure: just this patch would suffice to fix the issue on rtl83xx, or does one need the patched netifd as well? If I understand correctly netifd needs patching as well to make sure the issue itself is tackled completely, but this UCI defaults patch suffices to work around the issue in rtl83xx right now?

Or am I understanding it wrong? I will be testing it on 23.05.

**Comment by janh** [comment]
> To be sure: just this patch would suffice to fix the issue on rtl83xx, or does one need the patched netifd as well? If I understand correctly netifd needs patching as well to make sure the issue itself is tackled completely, but this UCI defaults patch suffices to work around the issue in rtl83xx right now?

The patched netifd is definitely needed, because it currently only understands the variant `network-device`.

If only the change to `ucidef_set_network_device_mac` in `uci-defaults.sh` is applied, then it would never work. Currently the issue only exists if `board.json` is modified again by another board.d script after `02_network` (because the key is changed from `network-device` to `network_device` by jshn while reading the data).

On the other hand, changing just netifd to accept `network_device` would actually fix it for rtl83xx. If netifd is changed to accept both variants (as the pull request does), then it should keep working in all cases. But for consistency it makes sense to change `ucidef_set_network_device_mac` in any case.

**Comment by MadnessASAP** [comment]
@Borromini @janh This PR alone is all that is needed, it includes a patch that will automatically be applied to netifd during the build.

The other PR against netifd is to upstream that patch.

**Comment by Borromini** [comment]
This won't get merged with the netifd patch included, but I suppose you'll update the PR once the netifd patch gets merged.

Tested-by: Stijn Segers <foss@volatilesystems.org>

**Comment by MadnessASAP** [comment]
> This won't get merged with the netifd patch included, but I suppose you'll update the PR once the netifd patch gets merged.

Once netifd gets patched, this PR practically becomes a noop, but yes.

Out of curiosity why wouldn't this get merged with the netifd patch? Is it just because patching a project under the openwrt umbrella and used (AFAIK) only by openwrt is a little weird?

**Comment by Borromini** [comment]
AFAIK patches should be sent in against those projects directly,

-------- Oorspronkelijk bericht --------
Op 8 okt. 2023 21:24, schreef MadnessASAP :

File: package/network/config/netifd/patches/110-config_network_device_fix.patch [hunk 1]
```diff
@@ -0,0 +1,55 @@
+From: Michael 'ASAP' Weinrich <michael@a5ap.net>
+Date: Thu, 5 Oct 2023 20:04:27 -0700
+Subject: [PATCH] config: netword-device is incompatible with jshn
+
+An oversight in openwrts uci-defaults.sh ucidef_set_network_device_*
+functions resulted in a hyphen (-) being used insted of a underscore (_)
```
Comment (raenye) [inline package/network/config/netifd/patches/110-config_network_device_fix.patch:6]: instead->instead
a underscore->an underscore

**Comment by Ansuel** [comment]
Also add right up the sob tag this tag

`Fixes: 9290539ca9c7 ("base-files: allow setting device and bridge macs")`

**Comment by MadnessASAP** [comment]
@Ansuel will do! Thank you. The patch was only in case this got merged before the netifd fix anyways.

**Comment by Ansuel** [comment]
@MadnessASAP any news with this? would really love to fix this ASAP

**Comment by Ansuel** [comment]
@svanheule I checked 1-2 thing yesterday and I will take care of fixing the thing in netifd (by using the _ value directly and not supporting both as I agree with you that it would be wrong) (I checked and this change can be done with no migration needed)

**Comment by svanheule** [comment]
Perhaps also worth mentioning (in the commit message) that this issue was exposed by commit 4ebba8a05d0, where the `board_config_load` call from 03_gpio induced the key normalization by jshn. See https://forum.openwrt.org/t/57875/2589

**Comment by MadnessASAP** [comment]
Dropped the patch and added the Fixes and Tested-By. Sorry for the delay was distracted by rather more petroleum based PRs.

**Comment by Ansuel** [comment]
@MadnessASAP can you squash them? Or if you give me permission I can take care of squashing them!

**Comment by MadnessASAP** [comment]
@Ansuel Squish!

**Comment by Ansuel** [comment]
@svanheule merged and backported.

@MadnessASAP Thanks a lot for pointing this out and proposed some fix!

**Comment by svanheule** [comment]
Thanks everyone!

**Pull request merged**